Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,826 @@
+
+    case (minfo, p) of
+        (Nothing, _)            -> return Nothing
+        (_, Nothing)            -> return Nothing
+        (Just minfo', Just p')  -> return $ Just (map (showSDocForUser dflags ghcQualify . ppr) $ modInfoExports minfo', p')
+
+type FullyQualifiedName = String    -- ^ e.g. e.g. "base-4.8.2.0:Data.Foldable.length"
+type StrModuleName      = String    -- ^ e.g. "Data.List"
+
+data MySymbol = MySymbolSysQualified  String  -- ^ e.g. "base-4.8.2.0:Data.Foldable.length"
+              | MySymbolUserQualified String  -- ^ e.g. "DL.length" with an import earlier like "import qualified Data.List as DL"
+              deriving Show
+
+data ModuleExports = ModuleExports
+    { mName            :: StrModuleName            -- ^ e.g. "Data.List"
+    , mPackageName     :: String                   -- ^ e.g. "snap-0.14.0.6"
+    , mInfo            :: NiceImportDecl           -- ^ Our parse of the module import, with info like "hiding (map)".
+    , qualifiedExports :: [FullyQualifiedName]     -- ^ e.g. [ "base-4.8.2.0:GHC.Base.++"
+                                                   --        , "base-4.8.2.0:GHC.List.filter"
+                                                   --        , "base-4.8.2.0:GHC.List.zip"
+                                                   --        , ...
+                                                   --        ]
+    }
+    deriving Show
+
+refineAs :: MySymbol -> [ModuleExports] -> [ModuleExports]
+
+-- User qualified the symbol, so we can filter out anything that doesn't have a matching 'modImportedAs'.
+refineAs (MySymbolUserQualified userQualSym) exports = filter f exports
+  where
+    f export = case modas of
+                Nothing     -> False
+                Just modas' -> modas' == userQualAs
+       where modas = modImportedAs $ mInfo export :: Maybe String
+
+             -- e.g. "DL"
+             userQualAs = fromMaybe (throw $ GMEString $ "ImportedFrom: expected a qualified name like 'DL.length' but got: " ++ userQualSym)
+                                    (moduleOfQualifiedName userQualSym)
+
+-- User didn't qualify the symbol, so we have the full system qualified thing, so do nothing here.
+refineAs (MySymbolSysQualified _) exports = exports
+
+refineRemoveHiding :: [ModuleExports] -> [ModuleExports]
+refineRemoveHiding exports = map (\e -> e { qualifiedExports = f e }) exports
+  where
+    f export = filter (`notElem` hiding') thisExports
+       where hiding = modHiding $ mInfo export :: [String] -- Things that this module hides.
+             hiding' = map (qualifyName thisExports) hiding  :: [String]    -- Qualified version of hiding.
+             thisExports = qualifiedExports export         -- Things that this module exports.
+
+    qualifyName :: [QualifiedName] -> String -> QualifiedName
+    qualifyName qualifiedNames name
+        -- = case filter (postfixMatch name) qualifiedNames of
+        = case nub (filter (name `f`) qualifiedNames) of
```

Comment:
Are you sure you need nub? `Set.toList . Set.fromList` would be `O(n log n)` instead of `O(n^2)`, is `n` likely to be large here?